src/components/AR/frameworks/xr8/xr8-provider.ts

```diff
@@ -0,0 +1,543 @@
+import * as QRCode from 'qrcode-svg';
+import { ARProvider } from '../../AR-provider.js';
+
+/**
+ * Array of extra permissions which some tracking mode might need. By default XR8 will need camera/microphone permissions and deviceMotion permission (iOS only). VPS for example must pass an extra 'location' permission
+ */
+export type XR8ExtraPermissions = Array<'location'>;
+
+
+interface XR8UIHandler {
+    requestUserInteraction: () => Promise<void>;
+    handlePermissionFail: (error: Error) => void;
+    handleError: (error: Event) => void;
+    showWaitingForDeviceLocation: () => void;
+    hideWaitingForDeviceLocation: () => void;
+    handleIncompatibleDevice: () => void;
+}
+
+class XR8Provider extends ARProvider {
+
+    public uiHandler: XR8UIHandler = new DefaultUIHandler();
+
+    public cachedWebGLContext: WebGL2RenderingContext | null = null;
+
+    // Loading of 8th Wall might be initiated by several components, make sure we load it only once
+    private _loading = false;
+
+    // XR8 currently provides no way to check if the session is running, only if the session is paused (and we never pause, we just XR8.end()). so we track this manually
+    private _running = false;
+
+    // Enforce the singleton pattern
+    private _instance: XR8Provider | null = null;
+
+    constructor() {
+        super();
+
+        // Safeguard that we are not running inside the editor
+        if (typeof document === 'undefined') {
+            return;
+        }
+
+        if (this._instance !== null) {
+            throw 'WebXRProvider cannot be instantiated';
+        }
+
+        this._instance = this;
+    }
+
+    public async load() {
```

Comment:
I would prefer that `XR8Provider` is only available once this finishes, that will also make your code a lot cleaner and avoid constantly checking for `_loading`.

If there are stuff that can be configured, they could be done in another object or something

Reply:
How that would work in case when multiple providers needs to be loaded?
consider this:
AR-face-tracking-camera component loads the XR8Provider:
```
const xr8Provider = await loadXR8Provider(); // <-- loads 5MB or so
ARSession.registerTrackingProvider(this.engine, xr8Provider);
```
in parallel AR-SLAM-camera component would load a native webXR provider:
 ```
 const webXRProvider = await loadWebXRProvider() // <-- actually nothing to load, will resolve instantelly
 ARSession.registerTrackingProvider(this.engine, webXRProvider);
 ```
 
 ARSession would register the `webXRProvider` way before it would register `xr8Provider` and would fire the `onARSessionReady` without even knowing that `xr8Provider` is not ready yet.
 
One alternative could pass a loader function to `this.provider = ARSession.loadTrackingProvider(this.engine, loadXR8Provider)`, but this way ARSession would still need to track if all providers are loaded and arguably this won't make code more readably.

Any suggestions how to solve it differently?

Reply:
You don't need to wait for it, you could actually do:

```js
let xr8Provider = null;
loadXR8Provider().then((provider) => xr8Provider = provider;)
```

So they would all run in parralel. Obviously, it would still be nice to have a system to tell the user when that's done or something, but it looks doable